File: 18.snailfish/hs/part2.hs

import Data.List.Split
import Data.Char

data Number = Int Int | Pair (Number, Number) deriving (Show)

parseHelper ('[':xs) = (Pair (first, second), tail rest)
    where
        (first, rest') = parseHelper xs
        (second, rest) = parseHelper (tail rest')
parseHelper string = (Int (read number), rest)
    where (number, rest) = span isDigit string

parse string = number
    where (number, _) = parseHelper string

applyFirst (Int x) dx = (Int (x + dx), 0)
applyFirst (Pair (x, y)) d = (Pair (x', y'), d'')
    where
        (x', d') = applyFirst x d
        (y', d'') = applyFirst y d'

applyLast (Int x) dx = (Int (x + dx), 0)
applyLast (Pair (x, y)) d = (Pair (x', y'), d'')
    where
        (y', d') = applyLast y d
        (x', d'') = applyLast x d'

magn (Int i) = i
magn (Pair (x, y)) = (3 * magn x) + (2 * magn y)

explodeStep (Int i) _ = (False, Int i, 0, 0)
explodeStep (Pair (Int l, Int r)) d
    | d >= 4 = (True, Int 0, l, r)
explodeStep (Pair (l, r)) d
    | b = (True, Pair (x, r'), dl, ddr')
    | b' = (True, Pair (l', x'), ddl', dr')
    | otherwise = (False, Pair (l, r), 0, 0)
    where
        (b, x, dl, dr) = explodeStep l (d + 1)
        (b', x', dl', dr') = explodeStep r (d + 1)
        (r', ddr') = applyFirst r dr
        (l', ddl') = applyLast l dl'

splitNumber i
    | even i = (Int (i `div` 2), Int (i `div` 2))
    | otherwise = (Int (i `div` 2), Int ((i `div` 2) + 1))

splitStep (Int i)
    | i >= 10 = (True, Pair (splitNumber i))
    | otherwise = (False, Int i)
splitStep (Pair (l, r))
    | b = (True, Pair (l', r))
    | b' = (True, Pair (l, r'))
    | otherwise = (False, Pair (l, r))
    where
        (b, l') = splitStep l
        (b', r') = splitStep r

reduceStep x
    | r = (True, x')
    | r' = (True, x'')
    | otherwise = (False, x)
    where
        (r, x', _, _) = explodeStep x 0
        (r', x'') = splitStep x

reduce x = if r then reduce x' else x
    where (r, x') = reduceStep x

add a b = reduce (Pair (a, b))

main = do
    dat <- getContents
    let numbers = map parse (filter (not . null) (splitOn "\n" dat)) :: [Number]
    let mmax = foldl max 0 [magn (add x y) | x <- numbers, y <- numbers]
    putStrLn ("Result: " ++ show mmax)

File: 18.snailfish/py/part2.py

import sys
import json

from functools import reduce

def flatten(number, d = 0):
    if isinstance(number, list):
        return flatten(number[0], d + 1) + flatten(number[1], d + 1)
    else:
        return [(number, d)]

numbers = [flatten(json.loads(s)) for s in sys.stdin.read().strip().split('\n')]

def expode(number):
    for i, (_, d) in enumerate(number):
        if d > 4:
            new = number[:i] + number[i + 1:]
            if i > 0:
                new[i - 1] = (new[i - 1][0] + number[i][0], new[i - 1][1])
            if i < len(new) - 1:
                new[i + 1] = (new[i + 1][0] + number[i + 1][0], new[i + 1][1])
            new[i] = (0, d - 1)
            return True, new
    return False, number

def split(number):
    for i, (n, d) in enumerate(number):
        if n >= 10:
            return True, number[:i] + [(n // 2, d + 1), ((n + 1) // 2, d + 1)] + number[i + 1:]
    return False, number

def normalize(number):
    while True:
        change, number = expode(number)
        if change:
            continue
        change, number = split(number)
        if not change:
            break
    return number

def add(a, b):
    return normalize([(x, d + 1) for x, d in a + b])

def magnitude(number):
    stack = []
    for v in number:
        stack.append(v)
        while len(stack) >= 2 and stack[-1][1] == stack[-2][1]:
            stack[-2] = (3 * stack[-2][0] + 2 * stack[-1][0], stack[-1][1] - 1)
            stack.pop()
    return stack[-1][0]

print('Result:', max(magnitude(add(x, y)) for x in numbers for y in numbers if x != y))

